### PR TITLE
todo/week-6

### PR DESCRIPTION
**Pull request by sghourab**
Completed SwiftUI animations course and updated Build a SwiftUI app for iOS 15

### CODE REVIEW

**Review by yura-hb — CHANGES_REQUESTED** [review]
@sghourab Hi. Could you please check, if you have pushed everything to the git, there are several files, which are missing (Topic.swift, Info.plist, schema.json). Also, you can check xcodegen, it is quite simple to install and you don't need to push xcodeproj

File: CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Model/Coin.swift [hunk 1]
```diff
@@ -0,0 +1,33 @@
+//
+//  Coin.swift
+//  CourseManager
+//
+//  Created by Summer Crow on 2022-11-11.
+//
+
+import Foundation
+
+struct Coin: Identifiable, Decodable {
```
Comment (yura-hb) [inline CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Model/Coin.swift:10]: It is better to make such models immutable, so you define

```
let id: Int
```

File: CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Model/Coin.swift [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    var id: Int
+    var coin_name: String
```
Comment (yura-hb) [inline CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Model/Coin.swift]: Use always camelCase

File: CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Model/Coin.swift [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    var acronym: String
+    var logo: String
+}
+
+class CoinModel: ObservableObject {
+    @Published var coins: [Coin] = []
```
Comment (yura-hb) [inline CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Model/Coin.swift]: It is better to make them `private(set)` so nobody can't modify them from the View.

File: CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Detail/CourseView.swift [hunk 4]
```diff
@@ -197,9 +207,9 @@ struct CourseView: View {
     }
 
     func fadeOut() {
-        appear[0] = false
-        appear[1] = false
-        appear[2] = false
+        for i in 0 ..< appear.count {
```
Comment (yura-hb) [inline CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Detail/CourseView.swift]: It can be written in shorter way
```
appear = appear.map { false }

appear = Array(repeating: false, count: appear.count)
```

File: CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Main/HomeView.swift [hunk 5]
```diff
@@ -8,30 +8,28 @@
...
-    @State var hasScrolled = false
     @Namespace var namespace
     @State var show = false
     @State var showStatusBar = true
     @State var selectedID = UUID()
     @State var showCourse = false
     @State var selectedIndex = 0
     @EnvironmentObject var model: Model
-    
+    @AppStorage("isLiteMode") var isLiteMode = true
```
Comment (yura-hb) [inline CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Main/HomeView.swift:18]: What is `LiteMode`?

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Utils/ExtensionsWeather.swift [hunk 6]
```diff
@@ -0,0 +1,37 @@
+//
```
Comment (yura-hb) [inline CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Utils/ExtensionsWeather.swift]: It is better to name file Weather + Extensions. Additionally, if you use colors only in weather, it is better to either add computed property/function directly to weather object

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Utils/ExtensionsWeather.swift [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    static let bottomSheetBorderTop = LinearGradient(gradient: Gradient(colors: [.white.opacity(0), .white.opacity(0.5), .white.opacity(0)]), startPoint: .leading, endPoint: .trailing)
+    static let underline = LinearGradient(gradient: Gradient(colors: [.white.opacity(0), .white, .white.opacity(0)]), startPoint: .leading, endPoint: .trailing)
+    static let tabBarBorder = Color("Weather Tab Bar Border").opacity(0.5)
+    static let forecastCardBackground = Color("Weather Forecast Card Background")
+    static let probabilityText = Color("Weather Probability Text")
+}
+
+extension View {
+    func innerShadow<S: Shape, SS: ShapeStyle>(shape: S, color: SS, lineWidth: CGFloat = 1, offsetX: CGFloat = 0, offsetY: CGFloat = 0, blur: CGFloat = 4, blendMode: BlendMode = .normal, opacity: Double = 1) -> some View {
+        return self
```
Comment (yura-hb) [inline CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Utils/ExtensionsWeather.swift]: return is unnecessary

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Utils/Shapes.swift [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+import Foundation
+import SwiftUI
+
+struct Arc: Shape {
+    func path(in rect: CGRect) -> Path {
+        var path = Path()
+
+              path.move(to: CGPoint(x: rect.minX - 1, y: rect.minY))
+              path.addQuadCurve(to: CGPoint(x: rect.maxX + 1, y: rect.minY), control: CGPoint(x: rect.midX, y: rect.midY))
```
Comment (yura-hb) [inline CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Utils/Shapes.swift]: Check alignment here

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Components/ForecastCard.swift [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+//  Created by Summer Crow on 2022-11-10.
+//
+
+import SwiftUI
+
+struct ForecastCard: View {
+    var forecast: Forecast
+    var forecastPeriod: ForecastPeriod
+    var isActive: Bool {
+        if forecastPeriod == ForecastPeriod.hourly {
```
Comment (yura-hb) [inline CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Components/ForecastCard.swift]: It is better to use `switch` statement here

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Components/ForecastCard.swift [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+            let isThisHour = Calendar.current.isDate(.now, equalTo: forecast.date, toGranularity: .hour)
```
Comment (yura-hb) [inline CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Components/ForecastCard.swift]: Here isThisHour is extra, because the method already tells the whole idea

File: CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Main/HomeView.swift (hunk 5 above)
Comment (Forkipe) [reply CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Main/HomeView.swift:18]: i think it makes app simpler ( removes animations and other stuff)

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Components/WeatherWidget.swift [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+struct WeatherWidget: View {
+    var forecast: Forecast
+    
+    var body: some View {
+        ZStack(alignment: .bottom) {
+            // MARK: Trapezoid
+            Trapezoid()
+                .fill(Color.weatherWidgetBackground)
+                .frame(width: 342, height: 174)
```
Comment (yura-hb) [inline CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Components/WeatherWidget.swift]: In almost all cases you can go without fixed frames, so try to update the code to note use them. The main reason is that the screen isn't adaptable to other devices screens

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Detail/WeatherView.swift [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+
+import SwiftUI
+
+struct WeatherView: View {
+    @State private var searchText = ""
+    
+    var searchResults: [Forecast] {
+        if searchText.isEmpty {
+            return Forecast.cities
+        } else {
```
Comment (yura-hb) [inline CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Detail/WeatherView.swift]: else here is extra, also ternary if statement looks shorter

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Main/HomeViewWeather.swift [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+                    .offset(y: -bottomSheetTranslationProrated * 46)
+
+                    BottomSheetView(position: $bottomSheetPosition) {
+                    } content: {
+                        ForecastView(bottomSheetTranslationProrated: bottomSheetTranslationProrated)
+                    }
+                    .onBottomSheetDrag { translation in
+                        bottomSheetTranslation = translation / screenHeight
+                        withAnimation(.easeInOut) {
+                               if bottomSheetPosition == BottomSheetPosition.top {
```
Comment (yura-hb) [inline CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Main/HomeViewWeather.swift]: hasDragged = bottomSheetPosition == .top?

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Main/HomeViewWeather.swift [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+                        bottomSheetPosition = .top
+                    })
+                    .offset(y: bottomSheetTranslationProrated * 115)
+
+                }
+            }
+            .navigationBarBackButtonHidden(true)
+        }
+    }
+    private var attributedString: AttributedString {
```
Comment (yura-hb) [inline CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Main/HomeViewWeather.swift]: Fix alignment here

Comment (yura-hb) [reply CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Main/HomeViewWeather.swift]: Also the point to consider, since the value is dependent only on the change of `hasDragged`, it may be a waste of resources to recompute it on every view update. Obviously, it isn't the issue here, but it may be in other place

File: CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Detail/SectionView.swift [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+        ZStack {
+            ScrollView {
+                cover
+                    .overlay(PlayView()
+                        .overlay(CircularView(value: section.progress, lineWidth: 5).padding(24)))
+                content
+                    .offset(y: 120)
+                    .padding(.bottom, 200)
+            }
+            .background(Color("Background"))
```
Comment (den-ree) [inline CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Detail/SectionView.swift]: @sghourab create please `enum AppColors` and put there all colours which you use in the app, instead of hardcoding them

File: CourseManager/Model/CourseInformation.swift [hunk 16]
```diff
@@ -17,8 +17,7 @@ struct CourseInformation: Identifiable, Codable {
     var status: Status
     var dateOfCompletion: Date
     var comments: String
-    
-    
+    //var destinationView: AnyView
```
Comment (den-ree) [inline CourseManager/Model/CourseInformation.swift]: remove the code

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Main/HomeViewWeather.swift [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+//  CourseManager
+//
+//  Created by Summer Crow on 2022-11-09.
+//
+
+import SwiftUI
+import BottomSheet
+
+enum BottomSheetPosition: CGFloat, CaseIterable {
+    case top = 0.83
```
Comment (den-ree) [inline CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Main/HomeViewWeather.swift]: Good that you use enum, and better to add comments why this values are used

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Detail/ForecastView.swift [hunk 18]
```diff
@@ -0,0 +1,72 @@
...
+                .frame(maxHeight: .infinity, alignment: .top)
+        }
+    }
+}
+
+struct ForecastView_Previews: PreviewProvider {
+    static var previews: some View {
+        ForecastView()
+            .background(Color.weatherBackground)
+            //.preferredColorScheme(.dark)
```
Comment (den-ree) [inline CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Detail/ForecastView.swift]: clean code.

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Components/WeatherWidget.swift [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+            
+            // MARK: Content
+            HStack(alignment: .bottom) {
+                VStack(alignment: .leading, spacing: 8) {
+                    // MARK: Forecast Temperature
```
Comment (den-ree) [inline CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Components/WeatherWidget.swift]: you don't need to use `// MARK:` to split up the views, instead of it create a new variable to explain type of view. 

`//Mark` should be used to split up group of methods or extensions not just small part of views.

File: CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Detail/CourseView.swift [hunk 20]
```diff
@@ -66,6 +68,7 @@ struct CourseView: View {
                     .frame(maxWidth: 500)
                     .matchedGeometryEffect(id: "image\(course.id)", in: namespace)
                     .offset(y: scrollY > 0 ? scrollY * -0.8 : 0)
+                    .accessibilityLabel("Cover Image")
```
Comment (den-ree) [inline CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Detail/CourseView.swift:71]: why did you add this?

File: CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Model/Handbook.swift [hunk 21]
```diff
@@ -0,0 +1,24 @@
...
+import SwiftUI
+
+struct Handbook: Identifiable {
+    let id = UUID()
+    var title: String
+    var subtitle: String
+    var text: String
+    var logo: String
+    var image: String
+    var color1: Color
```
Comment (den-ree) [inline CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Model/Handbook.swift:17]: what is the difference between color1 and color2 ?

**Review by den-ree — COMMENTED** [review]
@sghourab check comments and also do you use the code linting tool?

File: CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Detail/CourseView.swift (hunk 4 above)
Comment (sghourab) [reply CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Detail/CourseView.swift]: I've updated to appear = Array(repeating: false, count: appear.count). Thank you, spent a while trying to figure out better code for this earlier, wasn't aware of this.

File: CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Main/HomeView.swift (hunk 5 above)
Comment (sghourab) [reply CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Main/HomeView.swift:18]: yes exactly, it removes the blur and shadows to improve performance in particularly older phones or very big app with lots of components/views

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Utils/ExtensionsWeather.swift (hunk 7 above)
Comment (sghourab) [reply CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Utils/ExtensionsWeather.swift]: updated

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Utils/Shapes.swift (hunk 8 above)
Comment (sghourab) [reply CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Utils/Shapes.swift]: updated

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Components/ForecastCard.swift (hunk 9 above)
Comment (sghourab) [reply CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Components/ForecastCard.swift]: Oh yes of course. Updated to this: 

   var isActive: Bool {
        switch forecastPeriod {
        case .hourly:
            return Calendar.current.isDate(.now, equalTo: forecast.date, toGranularity: .hour)
        case .daily:
            return Calendar.current.isDate(.now, equalTo: forecast.date, toGranularity: .day)
        }
    }

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Components/ForecastCard.swift (hunk 10 above)
Comment (sghourab) [reply CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Components/ForecastCard.swift]: agreed, updated

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Components/WeatherWidget.swift (hunk 11 above)
Comment (sghourab) [reply CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Components/WeatherWidget.swift]: Completely agree. I had the same thoughts as I was following the instructor (who I believe was trying to make an exact copy of the figma design). I've updated to just include padding and looks perfectly fine.

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Utils/ExtensionsWeather.swift (hunk 6 above)
Comment (sghourab) [reply CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Utils/ExtensionsWeather.swift]: Hi Yura, I've updated the name of the file. Just want to double check I understood the second part correctly. I'm using the colors across the weather app and not changing any colors based on changes in the weather or forecast objects. Should I still be putting the colors as computed properties/functions in the weather object?

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Detail/WeatherView.swift (hunk 12 above)
Comment (sghourab) [reply CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Detail/WeatherView.swift]: Updated to the following: 
  var searchResults: [Forecast] {
        return searchText.isEmpty ? Forecast.cities : Forecast.cities.filter { $0.location.contains(searchText)
        }
    }

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Main/HomeViewWeather.swift (hunk 17 above)
Comment (sghourab) [reply CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Main/HomeViewWeather.swift]: This enum was used because a library for a pull up bottom sheet was included in the project. It uses this enum to know the position of the bottom sheet. You can define your own relative values or a default value is set. The instructor added her own values based on a figma file. I've updated the code as: 

enum BottomSheetPosition: CGFloat, CaseIterable {
    case top = 0.83 // 702/844 = position of sheet in 'top' position / overall height of screen
    case middle = 0.385 //325/844 = position of sheet in 'middle' position / overall height of screen
}

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Main/HomeViewWeather.swift (hunk 13 above)
Comment (sghourab) [reply CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Main/HomeViewWeather.swift]: I've updated this to: 
bottomSheetPosition == .top ? (hasDragged = true) : (hasDragged = false)

Comment (yura-hb) [reply CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Main/HomeViewWeather.swift]: It isn't the best though

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Main/HomeViewWeather.swift (hunk 17 above)
Comment (yura-hb) [reply CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Main/HomeViewWeather.swift]: @sghourab Just the point about such values. In general, you always want system values

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Utils/ExtensionsWeather.swift (hunk 6 above)
Comment (yura-hb) [reply CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Utils/ExtensionsWeather.swift]: Hi, it is better, that the weather will have `state` variable and which will provide color. For example, `state` is enum and it has associated color

File: CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Model/Coin.swift (hunk 1 above)
Comment (sghourab) [reply CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Model/Coin.swift:10]: updated

File: CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Model/Coin.swift (hunk 2 above)
Comment (sghourab) [reply CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Model/Coin.swift]: This is just embarrassing. I was blindly following the instructor, would never do this in my own app. I've updated to include CodingKeys

File: CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Model/Coin.swift (hunk 3 above)
Comment (sghourab) [reply CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Model/Coin.swift]: Wasn't aware of this, good to know. I've updated but also made the fetch func private as follows. Is this a good way to go about it?

 @Published private(set) var coins: [Coin] = []
   init() {
        Task(priority: .high) {
            await fetchCoins()
        }
    }
    @MainActor
    private func fetchCoins() async {
        do {
            guard let url = URL(string: "https://random-data-api.com/api/crypto_coin/random_crypto_coin?size=10") else {
                return
            }
            let (data, _) = try await URLSession.shared.data(from: url)
            coins = try JSONDecoder().decode([Coin].self, from: data)
        } catch(let error) {
            print(error)
        }
    }

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Main/HomeViewWeather.swift (hunk 13 above)
Comment (sghourab) [reply CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Main/HomeViewWeather.swift]: Oh sorry, you mean this: 
hasDragged = bottomSheetPosition == .top ? true : false

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Main/HomeViewWeather.swift (hunk 14 above)
Comment (sghourab) [reply CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Main/HomeViewWeather.swift]: Ok got it, will keep in mind.

File: CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Model/Handbook.swift (hunk 21 above)
Comment (sghourab) [reply CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Model/Handbook.swift:17]: The 2 colors are used to make a linear gradient color for the background of the handbook component at the bottom of this view. Do you think there should be a better way to name them?

Simulator Screen Shot - iPhone 14 Pro - 2022-11-15 at 12 11 01

File: CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Detail/CourseView.swift (hunk 20 above)
Comment (sghourab) [reply CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Detail/CourseView.swift:71]: At the end of the course, the instructor touched on accessibility and it probably should read something more descriptive than "Cover Image".  It's just an example of what can be done

File: CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Detail/SectionView.swift (hunk 15 above)
Comment (sghourab) [reply CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Detail/SectionView.swift]: Hi Den, I've made the enum AppColors, but wanted to get your opinion on making an extension to Color instead because it would end up in more concise code: 

enum AppColors: String {
    case background = "Background"
    case shadow = "Shadow"
}
when called: 
Color(AppColors.background.rawValue)

vs 
extension Color {
 static let background = Color("Background")
}

when called: 
Color.background

Comment (sghourab) [reply CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Detail/SectionView.swift:25]: Never mind, thought about it a bit more and will use this approach: 
enum MyColor {
    case color(red: CGFloat, green: CGFloat, blue: CGFloat, alpha: CGFloat)

    static let blue = Color(red: 0.29, green: 0.47, blue: 0.647)
    static let white = Color(red: 0.98, green: 0.98, blue: 1.0)
}

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Components/WeatherWidget.swift (hunk 19 above)
Comment (sghourab) [reply CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Components/WeatherWidget.swift]: This has been updated. I left them there as a reminder to where  I need to refactor to create variable of view

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Main/HomeViewWeather.swift (hunk 17 above)
Comment (sghourab) [reply CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Views/Main/HomeViewWeather.swift]: I'll keep that in mind for sure

File: CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Utils/ExtensionsWeather.swift (hunk 6 above)
Comment (sghourab) [reply CourseManager/Courses/Course 4 UI and Animations in SwiftUI/Utils/ExtensionsWeather.swift]: Hi Yura, I'm sorry I'm not really getting this. I understand putting the colors in an enum, but the state of colors doesn't change, they're constant across the Weather app.

**Comment by sghourab** [comment]
@Den-Ree and @yura-hb Thank you very much for your reviews. I think I've worked on all of them, except I haven't installed xcodegen just yet and will push project again so all files are available.
This is a link of the screencast that shows the last course: https://drive.google.com/file/d/1ShxxeIhUElXcYsCD6JZxfknD1wwgCzGv/view?usp=share_link
Moving forward, will be focusing on my side of the app if there are no more issues

File: CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Model/Handbook.swift (hunk 21 above)
Comment (den-ree) [reply CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Model/Handbook.swift:17]: @sghourab yes, it should be colorLeft/right, colorTop/bottom depends on the direction.

File: CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Main/HomeView.swift (hunk 5 above)
Comment (den-ree) [reply CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Main/HomeView.swift:18]: @Forkipe @sghourab can you share documentation with me?

**Review by den-ree — APPROVED** [review]

**Comment by den-ree** [comment]
@sghourab I will merge the PR, but still there are a lot of changes which a mixed with images. better to split up commits and make different PRs when you are adding files and do changes in files.

File: CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Main/HomeView.swift (hunk 5 above)
Comment (sghourab) [reply CourseManager/Courses/Course 1 Build a SwiftUI app for iOS 15/Views/Main/HomeView.swift:18]: @Den-Ree It's not a library, it's a Bool stored in AppStorage that we used throughout the app with the custom Modifier 'ifMode' to alter some views: 

extension View {
@ViewBuilder func ifMode<Content: View>(_ condition: Bool, transform: (Self) -> Content) -> some View {
        if condition {
            transform(self)
        } else {
            self
        }
    }}

And this can then be applied on any view, example: 
Rectangle
              .ifMode(!isLiteMode, transform: { view in
            view
                .shadow(color: AppColors.shadow.opacity(0.5), radius: 30, x: 0, y: 15)
        })

OR simply like this:

 Section {
                    Toggle(isOn: $isLiteMode) {
                        Label("Lite Mode", systemImage: isLiteMode ? "tortoise" : "hare")
                    }
                }
